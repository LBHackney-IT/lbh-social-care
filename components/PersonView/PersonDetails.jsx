import { useState } from 'react';
import PropTypes from 'prop-types';

const PersonDetails = ({
  otherNames,
  ageContext,
  firstName,
  lastName,
  mosaicId,
  dateOfBirth,
  nhsNumber,
  firstLanguage,
  religion,
  dateOfDeath,
  sexualOrientation,
  gender,
  ethnicity,
  addressList,
  phoneNumber,
  email,
  preferredMethodOfContact,
  expandView = false,
}) => {
  const [expandDetails, setExpandDetails] = useState(false);
  const address = addressList?.find(
    ({ displayAddressFlag }) => displayAddressFlag === 'Y'
  );
  return (
    <>
      <div>
        <div className="lbh-table-header">
          <h2 className="govuk-fieldset__legend--m govuk-custom-text-color">
            {expandView ? `${firstName} ${lastName}` : 'PERSON DETAILS'}
          </h2>
          {expandView && (
            <div
              className="govuk-link"
              role="button"
              onClick={() => setExpandDetails(!expandDetails)}
            >
              {expandDetails ? 'Collapse' : 'Expand'} view
            </div>
          )}
        </div>
        <hr className="govuk-divider" />
        {(!expandView || expandDetails) && (
          <>
            <dl className="govuk-summary-list">
              {otherNames?.length > 0 && (
                <div className="govuk-summary-list__row">
                  <dt className="govuk-summary-list__key">Other Names</dt>
                  <dd className="govuk-summary-list__value">
                    <ul className="govuk-list">
                      {otherNames.map(({ firstName, lastName }) => (
                        <li key={otherNames}>
                          {firstName} {lastName}
                        </li>
                      ))}
                    </ul>
                  </dd>
                </div>
              )}
              <div className="govuk-summary-list__row">
                <dt className="govuk-summary-list__key">Person ID</dt>
                <dd className="govuk-summary-list__value">#{mosaicId}</dd>
              </div>
              {gender && (
                <div className="govuk-summary-list__row">
                  <dt className="govuk-summary-list__key">Gender</dt>
                  <dd className="govuk-summary-list__value">{gender}</dd>
                </div>
              )}
              {dateOfBirth && (
                <div className="govuk-summary-list__row">
                  <dt className="govuk-summary-list__key">Date of birth</dt>
                  <dd className="govuk-summary-list__value">
                    {new Date(dateOfBirth).toLocaleDateString('en-GB')}
                  </dd>
                </div>
              )}
              {dateOfDeath && (
                <div className="govuk-summary-list__row">
                  <dt className="govuk-summary-list__key">Date Of Death</dt>
                  <dd className="govuk-summary-list__value">
                    {new Date(dateOfDeath).toLocaleDateString('en-GB')}
                  </dd>
                </div>
              )}
              {ethnicity && (
                <div className="govuk-summary-list__row">
                  <dt className="govuk-summary-list__key">Ethnicity</dt>
                  <dd className="govuk-summary-list__value">{ethnicity}</dd>
                </div>
              )}
              {firstLanguage && (
                <div className="govuk-summary-list__row">
                  <dt className="govuk-summary-list__key">First Language </dt>
                  <dd className="govuk-summary-list__value">{firstLanguage}</dd>
                </div>
              )}
              {religion && (
                <div className="govuk-summary-list__row">
                  <dt className="govuk-summary-list__key">Religion </dt>
                  <dd className="govuk-summary-list__value">{religion}</dd>
                </div>
              )}
              {sexualOrientation && (
                <div className="govuk-summary-list__row">
                  <dt className="govuk-summary-list__key">
                    Sexual orientation
                  </dt>
                  <dd className="govuk-summary-list__value">
                    {sexualOrientation}
                  </dd>
                </div>
              )}
              {nhsNumber && (
                <div className="govuk-summary-list__row">
                  <dt className="govuk-summary-list__key">NHS number</dt>
                  <dd className="govuk-summary-list__value">{nhsNumber}</dd>
                </div>
              )}
              {address?.addressLine1 && (
                <div className="govuk-summary-list__row">
                  <dt className="govuk-summary-list__key">Address</dt>
                  <dd className="govuk-summary-list__value">
                    <p>{address.addressLine1}</p>
                  </dd>
                </div>
              )}
              {phoneNumber?.length > 0 && (
                <div className="govuk-summary-list__row">
                  <dt className="govuk-summary-list__key">Phone Number</dt>
                  <dd className="govuk-summary-list__value">
                    <ul className="govuk-list">
                      {phoneNumber.map(({ phoneNumber, phoneType }) => (
                        <li key={phoneNumber}>
                          {phoneNumber} - {phoneType}
                        </li>
                      ))}
                    </ul>
                  </dd>
                </div>
              )}
              {email && (
                <div className="govuk-summary-list__row">
                  <dt className="govuk-summary-list__key">Email </dt>
                  <dd className="govuk-summary-list__value">{email}</dd>
                </div>
              )}
              {preferredMethodOfContact && (
                <div className="govuk-summary-list__row">
                  <dt className="govuk-summary-list__key">
                    Preferred Method Of Contact
                  </dt>
                  <dd className="govuk-summary-list__value">
                    {preferredMethodOfContact}
                  </dd>
                </div>
              )}
              {ageContext && (
                <div className="govuk-summary-list__row">
                  <dt className="govuk-summary-list__key">Person type</dt>
                  <dd className="govuk-summary-list__value">
                    {ageContext === 'C' ? 'CFS' : 'ASC'}
                  </dd>
                </div>
              )}
            </dl>
          </>
        )}
      </div>
    </>
  );
};
PersonDetails.propTypes = {
  expandView: PropTypes.bool,
  firstName: PropTypes.string.isRequired,
  lastName: PropTypes.string.isRequired,
  mosaicId: PropTypes.string,
  dateOfBirth: PropTypes.string,
  nhsNumber: PropTypes.string,
  nationality: PropTypes.string,
  gender: PropTypes.string,
  addressList: PropTypes.arrayOf(PropTypes.shape({})),
  phoneNumber: PropTypes.arrayOf(PropTypes.shape({})),
};
export default PersonDetails;
