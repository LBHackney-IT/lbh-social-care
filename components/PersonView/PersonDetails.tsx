import { useState } from 'react';

import { getEthnicityName } from 'utils/person';
import { Resident } from 'types';

interface Props {
  person: Resident;
  expandView?: boolean;
}

const PersonDetails = ({
  person: {
    otherNames,
    contextFlag,
    firstName,
    lastName,
    id,
    dateOfBirth,
    nhsNumber,
    firstLanguage,
    religion,
    dateOfDeath,
    sexualOrientation,
    gender,
    address,
    ethnicity,
    phoneNumbers,
    emailAddress,
    preferredMethodOfContact,
  },
  expandView = false,
}: Props): React.ReactElement => {
  const [expandDetails, setExpandDetails] = useState(false);
  return (
    <div>
      <div className="lbh-table-header">
        <h2 className="govuk-fieldset__legend--m govuk-custom-text-color">
          {expandView ? `${firstName} ${lastName}` : 'PERSON DETAILS'}
        </h2>
        {expandView && (
          <div
            className="govuk-link"
            role="button"
            onClick={() => setExpandDetails(!expandDetails)}
          >
            {expandDetails ? 'Collapse' : 'Expand'} view
          </div>
        )}
      </div>
      <hr className="govuk-divider" />
      {(!expandView || expandDetails) && (
        <>
          <dl className="govuk-summary-list">
            {otherNames && otherNames?.length > 0 && (
              <div className="govuk-summary-list__row">
                <dt className="govuk-summary-list__key">Other Names</dt>
                <dd className="govuk-summary-list__value">
                  <ul className="govuk-list">
                    {otherNames.map(({ firstName, lastName }) => (
                      <li key={`${firstName} ${lastName}`}>
                        {firstName} {lastName}
                      </li>
                    ))}
                  </ul>
                </dd>
              </div>
            )}
            <div className="govuk-summary-list__row">
              <dt className="govuk-summary-list__key">Person ID</dt>
              <dd className="govuk-summary-list__value">#{id}</dd>
            </div>
            {gender && (
              <div className="govuk-summary-list__row">
                <dt className="govuk-summary-list__key">Gender</dt>
                <dd className="govuk-summary-list__value">{gender}</dd>
              </div>
            )}
            {dateOfBirth && (
              <div className="govuk-summary-list__row">
                <dt className="govuk-summary-list__key">Date of birth</dt>
                <dd className="govuk-summary-list__value">
                  {new Date(dateOfBirth).toLocaleDateString('en-GB')}
                </dd>
              </div>
            )}
            {dateOfDeath && (
              <div className="govuk-summary-list__row">
                <dt className="govuk-summary-list__key">Date Of Death</dt>
                <dd className="govuk-summary-list__value">
                  {new Date(dateOfDeath).toLocaleDateString('en-GB')}
                </dd>
              </div>
            )}
            {ethnicity && (
              <div className="govuk-summary-list__row">
                <dt className="govuk-summary-list__key">Ethnicity</dt>
                <dd className="govuk-summary-list__value">
                  {getEthnicityName(ethnicity)}
                </dd>
              </div>
            )}
            {firstLanguage && (
              <div className="govuk-summary-list__row">
                <dt className="govuk-summary-list__key">First Language </dt>
                <dd className="govuk-summary-list__value">{firstLanguage}</dd>
              </div>
            )}
            {religion && (
              <div className="govuk-summary-list__row">
                <dt className="govuk-summary-list__key">Religion </dt>
                <dd className="govuk-summary-list__value">{religion}</dd>
              </div>
            )}
            {sexualOrientation && (
              <div className="govuk-summary-list__row">
                <dt className="govuk-summary-list__key">Sexual orientation</dt>
                <dd className="govuk-summary-list__value">
                  {sexualOrientation}
                </dd>
              </div>
            )}
            {nhsNumber && (
              <div className="govuk-summary-list__row">
                <dt className="govuk-summary-list__key">NHS number</dt>
                <dd className="govuk-summary-list__value">{nhsNumber}</dd>
              </div>
            )}
            {address && (
              <div className="govuk-summary-list__row">
                <dt className="govuk-summary-list__key">Address</dt>
                <dd className="govuk-summary-list__value">
                  <div>{address.address}</div>
                  <div>{address.postcode}</div>
                </dd>
              </div>
            )}
            {phoneNumbers.length > 0 && (
              <div className="govuk-summary-list__row">
                <dt className="govuk-summary-list__key">Phone Number</dt>
                <dd className="govuk-summary-list__value">
                  <ul className="govuk-list">
                    {phoneNumbers.map(({ number, type }) => (
                      <li key={number}>
                        {number} - {type}
                      </li>
                    ))}
                  </ul>
                </dd>
              </div>
            )}
            {emailAddress && (
              <div className="govuk-summary-list__row">
                <dt className="govuk-summary-list__key">Email</dt>
                <dd className="govuk-summary-list__value">{emailAddress}</dd>
              </div>
            )}
            {preferredMethodOfContact && (
              <div className="govuk-summary-list__row">
                <dt className="govuk-summary-list__key">
                  Preferred Method Of Contact
                </dt>
                <dd className="govuk-summary-list__value">
                  {preferredMethodOfContact}
                </dd>
              </div>
            )}
            {contextFlag && (
              <div className="govuk-summary-list__row">
                <dt className="govuk-summary-list__key">Person type</dt>
                <dd className="govuk-summary-list__value">
                  {contextFlag === 'C' ? 'CFS' : 'ASC'}
                </dd>
              </div>
            )}
          </dl>
        </>
      )}
    </div>
  );
};

export default PersonDetails;
