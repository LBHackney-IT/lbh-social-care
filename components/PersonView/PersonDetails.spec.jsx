import { render, fireEvent } from '@testing-library/react';

import PersonDetails from './PersonDetails';

describe('PersonDetails component', () => {
  const props = {
    otherNames: [{ firstName: 'asd', laseName: 'qwe' }],
    ageContext: 'CFS',
    firstName: 'i am the first',
    lastName: 'i am the last',
    mosaicId: 'mosaic',
    dateOfBirth: '1978-02-23T00:00:00.0000000',
    dateOfDeath: '1998-02-23T00:00:00.0000000',
    nhsNumber: 'nhs_123',
    ethnicity: 'White',
    gender: 'X',
    sexualOrientation: 'straight',
    email: 'test@test.com',
    preferredMethodOfContact: 'email',
    addressList: [
      {
        contactAddressFlag: 'N',
        displayAddressFlag: 'N',
        addressLine1: 'old adress',
        addressLine2: null,
        addressLine3: null,
        postCode: 'E5 0PU',
      },
      {
        contactAddressFlag: 'N',
        displayAddressFlag: 'Y',
        addressLine1: 'new adress',
        addressLine2: null,
        addressLine3: null,
        postCode: 'E5 0PU',
      },
    ],
    phoneNumber: [
      {
        phoneNumber: '02123',
        phoneType: 'Home',
      },
    ],
  };

  it('should render properly', () => {
    const { asFragment } = render(<PersonDetails {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render properly - expanded', async () => {
    const { asFragment, getByRole } = render(
      <PersonDetails {...props} expandView />
    );
    expect(asFragment()).toMatchSnapshot();
    fireEvent.click(getByRole('button'));
    expect(asFragment()).toMatchSnapshot();
  });
});
