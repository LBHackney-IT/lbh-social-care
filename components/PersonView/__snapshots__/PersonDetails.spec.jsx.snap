// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonDetails component should render properly - expanded 1`] = `
<DocumentFragment>
  <div>
    <div
      class="lbh-table-header"
    >
      <h2
        class="govuk-fieldset__legend--m govuk-custom-text-color"
      >
        i am the first i am the last
      </h2>
      <div
        class="govuk-link"
        role="button"
      >
        Expand view
      </div>
    </div>
    <hr
      class="govuk-divider"
    />
  </div>
</DocumentFragment>
`;

exports[`PersonDetails component should render properly - expanded 2`] = `
<DocumentFragment>
  <div>
    <div
      class="lbh-table-header"
    >
      <h2
        class="govuk-fieldset__legend--m govuk-custom-text-color"
      >
        i am the first i am the last
      </h2>
      <div
        class="govuk-link"
        role="button"
      >
        Collapse view
      </div>
    </div>
    <hr
      class="govuk-divider"
    />
    <dl
      class="govuk-summary-list"
    >
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Other Names
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          test test
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Person ID
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          #mosaic
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Gender
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          X
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Date of birth
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          2/23/1978
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Date Of Death
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          1998-02-23T00:00:00.0000000
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Ethnicity
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          White
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Religion 
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          straight
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          NHS number
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          nhs_123
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Address
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          <p>
            new adress
          </p>
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Phone Number
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          <ul
            class="govuk-list"
          >
            <li>
              02123 - Home
            </li>
            <div
              class="govuk-summary-list__row"
            >
              <dt
                class="govuk-summary-list__key"
              >
                Email 
              </dt>
              <dd
                class="govuk-summary-list__value"
              >
                test@test.com
              </dd>
            </div>
            <div
              class="govuk-summary-list__row"
            >
              <dt
                class="govuk-summary-list__key"
              >
                Preferred Method Of Contact
              </dt>
              <dd
                class="govuk-summary-list__value"
              >
                email
              </dd>
            </div>
          </ul>
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Person type
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          ASC
        </dd>
      </div>
    </dl>
  </div>
</DocumentFragment>
`;

exports[`PersonDetails component should render properly 1`] = `
<DocumentFragment>
  <div>
    <div
      class="lbh-table-header"
    >
      <h2
        class="govuk-fieldset__legend--m govuk-custom-text-color"
      >
        PERSON DETAILS
      </h2>
    </div>
    <hr
      class="govuk-divider"
    />
    <dl
      class="govuk-summary-list"
    >
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Other Names
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          test test
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Person ID
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          #mosaic
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Gender
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          X
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Date of birth
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          2/23/1978
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Date Of Death
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          1998-02-23T00:00:00.0000000
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Ethnicity
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          White
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Religion 
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          straight
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          NHS number
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          nhs_123
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Address
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          <p>
            new adress
          </p>
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Phone Number
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          <ul
            class="govuk-list"
          >
            <li>
              02123 - Home
            </li>
            <div
              class="govuk-summary-list__row"
            >
              <dt
                class="govuk-summary-list__key"
              >
                Email 
              </dt>
              <dd
                class="govuk-summary-list__value"
              >
                test@test.com
              </dd>
            </div>
            <div
              class="govuk-summary-list__row"
            >
              <dt
                class="govuk-summary-list__key"
              >
                Preferred Method Of Contact
              </dt>
              <dd
                class="govuk-summary-list__value"
              >
                email
              </dd>
            </div>
          </ul>
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Person type
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          ASC
        </dd>
      </div>
    </dl>
  </div>
</DocumentFragment>
`;
